### DISCUSSION TITLE
zsh-bench

### DISCUSSION THREAD

**Discussion by romkatv** [starship/starship#3180] [General]
Hey, neighbours!

When you have a minute, please check out https://github.com/romkatv/zsh-bench. It's specifically for zsh but might be of use to you -- cross-shell folks -- too.

- It's something you can use to automate the measurement of various user-visible shell latencies.
- It's something you can ask your users to run when diagnosing slow shell.

The document mentions starship -- there is just one paragraph, easy to find -- and it's not flattering. Can you double-check that I haven't screwed up the setup? I install starship on ubuntu 21.10 [like this](https://github.com/romkatv/zsh-bench/blob/5425159141b1a0ad422f67c9d9df1b0a487e7a24/internal/install-deps#L20-L21), use [this zshrc](https://github.com/romkatv/zsh-bench/blob/bc5a2e73b1ff7ad47b2fc7ea57a53aa87d6a6acb/configs/starship/skel/.zshrc) and [explicitly disable](https://github.com/romkatv/zsh-bench/blob/bc5a2e73b1ff7ad47b2fc7ea57a53aa87d6a6acb/configs/starship/skel/.zshenv) all other rc files.

You can run the benchmark on any machine that has `docker` command. The instructions are on the homepage. You can also use the following command to create a playground in a docker container:

```zsh
curl -fsSLO https://raw.githubusercontent.com/romkatv/zsh-bench/master/playground
chmod +x ./playground
./playground starship
```

This playground has the same environment as during the benchmarking. When prompt appears, you'll be in a git repo with 1k directories and 10k files. Note that prompt is visibly lagging even though `git status` is fairly fast (94ms command lag vs 14ms `git status` wall time). It's slower than all other prompts I've benchmarked, which includes slow prompts -- by zsh standards -- robbyrussel (the default prompt in ohmyzsh), agnoster (the classic powerline prompt) and asciiship (the 7-bit contribution to the fleet). Is this expected? This is without a starship-specific config. Please look around in there to see if something looks off. Maybe it's because my machine has 64 cores? I haven't tried it on other machines, so other data points would be appreciated. Maybe starship is tripping over something? Maybe the container I've set up is missing something crucial or having something unusual? (My best guess is mentioned in the doc -- starship clones *a lot*, and clone is one of the most expensive syscalls you can make; I don't know *why* starship clones so much though; creating a ton of threads just to join them? might be easy to fix if so.)

You can run `~/zsh-bench/zsh-bench --iters 16` while in the container to benchmark login shell. Handy if you want to modify something in there and see the effect. Larger `--iters` will give you more accurate results but it takes longer to run. You can clone zsh-bench repo and run the same thing on your local machine to benchmark your personal login shell (as long as it's zsh).

If you need to install more stuff in the container, run `sudo apt install <pkg>`. The container disappears when you exit the shell.

If nothing looks off but you feel uncomfortable with my portrayal of starship, please let me know. I can remove it -- starship was a late addition to the list of benchmarked configs, it's not the focus of the doc and mentioning it doesn't contribute to the main points of the analysis (I don't extend the same offer to the devs of pure zsh projects mentioned in the doc). I hope you'll still be able to benefit from zsh-bench even if starship is not listed among benchmarked projects and I'll do my best to support your use cases.

If you've read this whole post before reading the homepage of [zsh-bench](https://github.com/romkatv/zsh-bench), please read it and then read the post again. I haven't done a very good job in this post (zsh-bench doc is decent though), so reading it twice might help with understanding my stumbling writing.

Edit: I've just rerun all benchmarks on ubuntu 21.10 (the original data was compiled on ubuntu 20.04) and starship looks a lot better. It still has large *command lag* but it's no longer the largest among the benchmarked themes. Perhaps that's the expected level of performance?

**Comment by matchai** [comment]
Hey Roman 👋 
Thank you for sharing. This looks excellent!
We'll strongly consider adding zsh-bench to our benchmarking pipeline as we get the ball rolling on the development of Starship v2.
I have no problems keeping Starship's benchmarks there. I'll follow-up with you when we get around to testing it ourselves, if the numbers don't line up.

**Reply by romkatv** [reply]
Sounds good!
